src/crab/ebpf_domain.cpp

```diff
@@ -1767,11 +1762,75 @@ void ebpf_domain_t::operator()(const Bin& bin) {
+                        uint64_t output = input << imm;
                         m_inv.set(dst.value, crab::interval_t{number_t{output}});
                         break;
                     }
+                } else if (interval.finite_size()) {
+                    number_t lb = interval.lb().number().value();
+                    number_t ub = interval.ub().number().value();
+                    if (bin.is64) {
+                        uint64_t lb_n = lb.cast_to_uint64() << imm;
+                        uint64_t ub_n = ub.cast_to_uint64() << imm;
+                        m_inv.set(dst.value, crab::interval_t{lb_n, ub_n});
+                    } else {
+                        number_t lb_w = lb.cast_to_finite_width(finite_width);
+                        number_t ub_w = ub.cast_to_finite_width(finite_width);
+                        uint64_t lb_n = lb_w.cast_to_uint32() << imm;
+                        uint64_t ub_n = ub_w.cast_to_uint32() << imm;
+                        if (lb_n > ub_n) {
+                            // Range is an int32_t, so use the full uint32_t range.
+                            m_inv.set(dst.value, crab::interval_t{0, UINT32_MAX});
+                        } else {
+                            // Range is contiguous as a uint32_t.
+                            m_inv.set(dst.value, crab::interval_t{lb_n, ub_n});
+                        }
+                    }
+                    break;
+                } else if (!bin.is64) {
+                    m_inv.set(dst.value, crab::interval_t{0, UINT32_MAX});
+                    break;
+                }
+            }
+            // avoid signedness and overflow issues in shl_overflow(dst.value, imm);
+            shl_overflow(dst.value, imm);
+            havoc_offsets(bin.dst);
+            break;
+        case Bin::Op::RSH:
+            if (m_inv.entail(type_is_number(bin.dst))) {
+                auto interval = m_inv.eval_interval(dst.value);
+                if (interval.finite_size()) {
+                    number_t lb = interval.lb().number().value();
+                    number_t ub = interval.ub().number().value();
+                    if (bin.is64) {
+                        uint64_t lb_n = lb.cast_to_uint64() >> imm;
+                        uint64_t ub_n = ub.cast_to_uint64() >> imm;
+                        m_inv.set(dst.value, crab::interval_t{lb_n, ub_n});
+                    } else {
+                        number_t lb_w = lb.cast_to_finite_width(finite_width);
+                        number_t ub_w = ub.cast_to_finite_width(finite_width);
+                        uint64_t lb_n = lb_w.cast_to_uint32() >> imm;
+                        uint64_t ub_n = ub_w.cast_to_uint32() >> imm;
+                        if (lb_n > ub_n) {
+                            // Range is an int32_t, so use the full uint32_t range.
+                            m_inv.set(dst.value, crab::interval_t{0, UINT32_MAX >> imm});
```

Comment:
Untested branch.

Reply:
Turns out this can't happen due to line 1713, so removed and added a comment.